source/common/network/transport_socket_options_impl.cc

```diff
@@ -58,50 +58,44 @@ void TransportSocketOptionsImpl::hashKey(std::vector<uint8_t>& key,
   commonHashKey(*this, key, factory);
 }
 
-TransportSocketOptionsConstSharedPtr
-TransportSocketOptionsUtility::fromFilterState(const StreamInfo::FilterState& filter_state) {
+TransportSocketOptionsConstSharedPtr TransportSocketOptionsUtility::fromFilterState(
+    const StreamInfo::FilterStateSharedPtr& filter_state) {
+  if (!filter_state) {
+    return nullptr;
+  }
   absl::string_view server_name;
   std::vector<std::string> application_protocols;
   std::vector<std::string> subject_alt_names;
   std::vector<std::string> alpn_fallback;
   absl::optional<Network::ProxyProtocolData> proxy_protocol_options;
 
-  bool needs_transport_socket_options = false;
```

Comment:
I could easily be misremembering but I thought previously this being null vs non-null meant that connections would or would not be hashed together in connection pools.  have you made sure it's safe to always return this?
cc @lizan who wrote this code

Reply:
Good point. I assumed this is just some optimization but if `nullptr` is actually special, then I'll have to track its usage throughout the code base.

Reply:
I looked over and I think there is no risk in returning an empty options struct vs `nullptr`. This is because the common hash iterates over the optionals and vectors to add to hash, and if they are empty, then it is effectively a no-op.

PS. I added some additional refactor to make it more clear the goal of this PR. I want to swap out `hashKey` definition site to the transport socket factory so that the factory can change the hash (rather that the options checking whether a factory is special).